src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,733 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+            std::mem::size_of::<T>(),
+            self.data.len(),
+            "Attempt to cast aarch64 register to incorrect type"
         );
 
-        let mut kvi: kvm_bindings::kvm_vcpu_init = kvm_bindings::kvm_vcpu_init::default();
-        vm.get_preferred_target(&mut kvi).unwrap();
-        vcpu.vcpu_init(&kvi).unwrap();
+        // # Safety:
+        // We checked that size of `T` is same as `data.len`
+        unsafe { &*(self.data.as_ptr().cast::<T>()) }
+    }
 
-        assert!(setup_boot_regs(&vcpu, 0, 0x0, &mem).is_ok());
+    /// Returns register data as a byte slice
+    pub fn as_slice(&self) -> &[u8] {
+        self.data
     }
-    #[test]
-    fn test_read_mpidr() {
-        let kvm = Kvm::new().unwrap();
-        let vm = kvm.create_vm().unwrap();
-        let vcpu = vm.create_vcpu(0).unwrap();
-        let mut kvi: kvm_bindings::kvm_vcpu_init = kvm_bindings::kvm_vcpu_init::default();
-        vm.get_preferred_target(&mut kvi).unwrap();
-
-        // Must fail when vcpu is not initialized yet.
-        let res = get_mpidr(&vcpu);
+}
+
+/// Reference to the aarch64 register.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Aarch64RegisterRefMut<'a> {
+    /// ID of the register
+    pub id: u64,
+    data: &'a mut [u8],
+}
+
+impl<'a> Aarch64RegisterRefMut<'a> {
+    /// Creates new register reference with provided id and data
+    pub fn new(id: u64, data: &'a mut [u8]) -> Self {
         assert_eq!(
-            res.unwrap_err(),
-            Error::GetOneReg(MPIDR_EL1, kvm_ioctls::Error::new(8))
+            reg_size(id) as usize,
+            data.len(),
+            "Attempt to create a register reference with incompatible id and data length"
         );
 
-        vcpu.vcpu_init(&kvi).unwrap();
-        assert_eq!(get_mpidr(&vcpu).unwrap(), 0x8000_0000);
+        Self { id, data }
     }
 
-    #[test]
-    fn test_get_set_regs() {
-        let kvm = Kvm::new().unwrap();
-        let vm = kvm.create_vm().unwrap();
-        let vcpu = vm.create_vcpu(0).unwrap();
-        let mut kvi: kvm_bindings::kvm_vcpu_init = kvm_bindings::kvm_vcpu_init::default();
-        vm.get_preferred_target(&mut kvi).unwrap();
-
-        // Must fail when vcpu is not initialized yet.
-        let mut regs = Vec::new();
-        let res = get_all_registers(&vcpu, &mut regs);
+    /// Returns register size in bytes
+    pub fn size(&self) -> RegSize {
+        reg_size(self.id).into()
+    }
+
+    /// Returns a reference to the register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn value<T>(&self) -> &T {
+        assert_eq!(
+            std::mem::size_of::<T>(),
+            self.data.len(),
+            "Attempt to cast aarch64 register to incorrect type"
+        );
+
+        // # Safety:
+        // We checked that size of `T` is same as `data.len`
+        unsafe { &*(self.data.as_ptr().cast::<T>()) }
+    }
```

Comment:
Mh, this needs to technically be marked as unsafe, as this is just an inlined version of `std::mem::transmute`, so all considerations of https://doc.rust-lang.org/std/mem/fn.transmute.html apply. 

There's a trait in rust-vmm/vmm-memory, `ByteValued` [[1]] which provides a safe interface for these kind of conversions. we should bound `T` by it and use its `from_slice` method to avoid unsoundness/unsafe code.

[[1]]: https://github.com/rust-vmm/vm-memory/blob/f93faaddc50933bc5f3739dbfbf8fec5dcd8e1d9/src/bytes.rs#L11

Reply:
So you propose to add `vm-memory` as a dependency to use one trait out of it which literally does same things, but now code is not int FC, but in that dependency?  I guess we need third opinion on this.

Reply:
Ah, whoops, forgot that we dont directly depend on vm-memory, but rather through the shim layer in utils (which is already a dependency of vmm) - sorry. We publically re-export the trait in `utils::vm_memory`.

Reply:
Ok, so we can use `ByteValued` for `RegData` to replace `as_slice` and `as_mut_slice`. But for `value<T>` this will not work, because `from_slice` requires slice to be aligned, and this is not guarantied in our case.

Reply:
Actually even using `ByteValued` for `RegData` is problematic, because it is not implemented for `u128` or any arrays that we use.

Reply:
Mh, right, but taking unaligned references is undefined behavior: https://doc.rust-lang.org/reference/behavior-considered-undefined.html, you'd need to use `std::ptr::read_unaligned`.

ByteValued not being implemented for u128 and arrays is indeed annoying though - maybe we could submit a PR that implements it for u/i128 and also turns the up-to-32 array impl into one that uses const generics? (although maybe as a follow up)

Reply:
Actually, some of the unit tests are even failing because of the unaligned reads :(

Reply:
Well, because all registers are stored in one continues memory, there is no way to avoid unaligned access.
Also `std::ptr::read_unaligned` is not what is needed here, because we need references.

Reply:
Mh, why do we definitely need a reference here? It looks like all use-sites of this function immediately dereference the returned reference, so returning an owned value here would not fundamentally break anything.

If you want to use a direct read from the memory inside the contiguous vec, you will have to use an unaligned read via `std::ptr::read_unaligned`. Otherwise, in debug mode you'll hit assertions, and in release mode you'll get a SIGILL because llvm will emit instructions that do not work for non-aligned accesses (iirc x86 processors do not care, but ARM does).

So the two possibility I see here are using `std::ptr::read_unaligned` and being very careful to never dereference pointers that are not aligned for their data types (and consequently dealing with raw pointers everywhere), or loosely a tiny bit of generality and using `{u16, u32, u64, u128}::from_ne_bytes` to go directly from slices to integral types (which I would honestly prefer, as it will eliminate all need for unsafe).

Reply:
Updated methods on `Aarch64RegisterRef` and `Aarch64RegisterRefMut` to return just a value with `read_unaligned` and added `set_value` for `Aarch64RegisterRefMut`. This does not eliminate `unsafe` fully, but guess it is better now. The remaining `unsafe` is mainly about casting types to slices and we can replace it with `ByteValued`, but it is would be literally same code.

Reply:
Mh, yes, this reduces the amount of unsafe and fixes the unaligned accesses. However, it is still essentially just https://doc.rust-lang.org/std/mem/fn.transmute_copy.html, so the function really should be marked as `unsafe`. Just checking for the size is not sufficient to make this safe. For example, one could try reading a u64 register into a `&mut u8` and create an invalid reference using this method, which is unsound.

Since this is only ever used with numeric types (and arrays I suppose, once we have registers wider than 128 bits), can this function instead be 

```rs
fn value<const N: usize>(&self) -> [u8; N] {
  assert_eq!(...);

  self.data.try_into().unwrap()
}
```
?

this way, its immediately compatible with https://doc.rust-lang.org/std/primitive.i128.html#method.from_le_bytes and related methods, and thus no unsafe is needed anywhere. I suppose if the extra function invocation at the call site is not desired, then this can be solved using some adhoc trait that realizes a `impl<const N: usize> TryFrom<[u8; N]> for {u8, u16, u32, u64, u128}`, which I think might be a better level of abstraction to choose here.

Reply:
Hmm... actually, I think marking all these functions as `unsafe` is not a bad idea. We only use these structs/methods internally, so situations like `ref.value::<&mut u8>()` are not a real possibility. This will create more `unsafe` blocks, but again, because we control everything, it is not a real issue.

Reply:
Ok, I adopted another approach, which does not need `unsafe`.